### PR TITLE
João Ludovico

### PR DESCRIPTION
**Pull request by jlouros**

### CODE REVIEW

**Comment by oranintelliflo** [comment]
- Overwrites the file whenever UploadFile endpoint is called
- Ask in the interview about logging when an exception is thrown
- Doesn't create new fund details for an existing fund (returns Conflict - check if they understood the requirements)
- When returning a failure result from a method there are no details of the error. For example: Returning bad request without indicating in which fund details the validation issue was
- Ask in the interview pros/cons of generalising the code around breakdowns. At the moment the code is explicit for each type of breakdown

File: scripts/Reset-Database.txt [hunk 1]
```diff
@@ -0,0 +1,28 @@
+Begin {
```
Comment (jlouros) [inline scripts/Reset-Database.txt]: why did this file got renamed?

File: src/Microservice.Api/Controllers/FundsController.cs [hunk 2]
```diff
@@ -34,15 +35,15 @@ public async Task<IEnumerable<FundDetails>> Get()
...
-                            // TO DO, code crashes if we include regional breakdown items
-                                // .ThenInclude(f => f.Items)
+                                 .ThenInclude(f => f.Items)
                             .Include(f => f.AssetBreakdown)
-                            // TO DO, code crashes if we include asset breakdown items
-                                // .ThenInclude(f => f.Items)
+                                 .ThenInclude(f => f.Items)
                             .ToListAsync().Result;
+
+            FundMapper fundMapper = new FundMapper();
```
Comment (jlouros) [inline src/Microservice.Api/Controllers/FundsController.cs]: did not use Dependency Injection

File: src/Microservice.Api/Controllers/FundsController.cs [hunk 3]
```diff
@@ -67,26 +70,63 @@ public IActionResult Post([FromBody] object payload)
...
+                    return Conflict("Fund with greater \"priceOn\" already exists");
+                }
+
+                UpdateFund(fund);
+                return StatusCode(200, "Record updated");
+            }
+            else
+            {
+                CreateFund(fund);
+                return Created("http://localhost:5000/api/funds", fundsDetails.id);
```
Comment (jlouros) [inline src/Microservice.Api/Controllers/FundsController.cs]: hardcoded location Url

File: src/Microservice.Api/Controllers/ImportController.cs [hunk 4]
```diff
@@ -27,31 +28,57 @@ public ImportController(FundsDbContext context)
...
+
         [HttpPost]
         [Consumes("application/zip")]
         public async Task<IActionResult> UploadFile()
         {
             using (var sr = new StreamReader(Request.Body))
             {
 
-                using (FileStream fs = new FileStream(Path.Combine(".", "asd.zip"), FileMode.CreateNew, FileAccess.Write))
+                using (FileStream fs = new FileStream(Path.Combine(".", "asd.zip"), FileMode.Create, FileAccess.Write))
```
Comment (jlouros) [inline src/Microservice.Api/Controllers/ImportController.cs]: name clash if multiple uploads are done in within the same second

File: src/Microservice.Api/Utils/FundValidations.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
+using Microservice.Api.Contracts;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+
+namespace Microservice.Api.Utils
+{
+    public static class FundValidations
```
Comment (jlouros) [inline src/Microservice.Api/Utils/FundValidations.cs]: why use of static methods?

**Comment by jlouros** [comment]
### This candidate was unable to solve the following issues:

- did not use dependency injection
- did not use logging
- partial unit test setup (did not fix previous tests)
- did not create for his own validation class
- strongly type configuration

**Pull request closed without merging**